src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/UnsignedAdapters.java

```diff
@@ -0,0 +1,159 @@
+            SHORT_TO_UNSIGNED_LONG = MethodHandles.lookup().findStatic(Short.class, "toUnsignedLong",
+                    MethodType.methodType(long.class, short.class));
+            LONG_TO_INT = MethodHandles.explicitCastArguments(MethodHandles.identity(int.class),
+                    MethodType.methodType(int.class, long.class));
+            INT_TO_UNSIGNED_LONG = MethodHandles.lookup().findStatic(Integer.class, "toUnsignedLong",
+                    MethodType.methodType(long.class, int.class));
+        } catch (Throwable ex) {
+            throw new ExceptionInInitializerError(ex);
+        }
+    }
+
+    public static VarHandle asUnsigned(VarHandle target, final Class<?> adaptedType) {
+        Objects.requireNonNull(target);
+        Objects.requireNonNull(adaptedType);
+        final Class<?> carrier = target.varType();
+        checkWidenable(carrier);
+        checkNarrowable(adaptedType);
+        checkTargetWiderThanCarrier(carrier, adaptedType);
+
+        if (adaptedType == int.class && carrier == byte.class) {
+            return intToUnsignedByte(target);
+        } else if (adaptedType == int.class && carrier == short.class) {
+            return intToUnsignedShort(target);
+        } else if (adaptedType == long.class && carrier == byte.class) {
+            return longToUnsignedByte(target);
+        } else if (adaptedType == long.class && carrier == short.class) {
+            return longToUnsignedShort(target);
+        } else if (adaptedType == long.class && carrier == int.class) {
+            return longToUnsignedInt(target);
+        } else {
+            throw new InternalError("should not reach here");
+        }
+    }
+
+    // int to byte
+    private static VarHandle intToUnsignedByte(VarHandle target) {
+        if (target.varType() != byte.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, INT_TO_BYTE, BYTE_TO_UNSIGNED_INT);
+    }
+
+    // int to short
+    private static VarHandle intToUnsignedShort(VarHandle target) {
+        if (target.varType() != short.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, INT_TO_SHORT, SHORT_TO_UNSIGNED_INT);
+    }
+
+    // long to byte
+    private static VarHandle longToUnsignedByte(VarHandle target) {
+        if (target.varType() != byte.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, LONG_TO_BYTE, BYTE_TO_UNSIGNED_LONG);
+    }
+
+    // long to short
+    private static VarHandle longToUnsignedShort(VarHandle target) {
+        if (target.varType() != short.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, LONG_TO_SHORT, SHORT_TO_UNSIGNED_LONG);
+    }
```

Comment:
I'm not super sure of the added value of these - the check is essentially a no op, since you already have a switch where, based on the var handle carrier, you decide which method to call. So perhaps we can just inline the right MH filter inside the switch above?